src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml

```diff
@@ -1426,6 +1434,21 @@
     <Field Name="Property" Type="PropertySymbol" Null="disallow"/>
     <Field Name="Index" Type="int"/>
   </Node>
+  <Node Name="BoundDagIndexerEvaluation" Base="BoundDagEvaluation">
+    <Field Name="IndexerType" Type="TypeSymbol"/>
+    <Field Name="LengthTemp" Type="BoundDagTemp"/>
```

Comment:
This usage of a temp implies a dependency between this evaluation and the evaluation of the length temp. I think this means we should update `TopologicallySortedNodes` so that this dependency is factored into the sorting.

Reply:
We operate on the decision dag nodes for sorting, rather than individual tests and evaluations, which means the sort order depends on the dag contruction algorithm - that's what we need to adjust to influence the sort order.

However, since we currently use the left-right heuristic for that, we never reorder tests. so for the following sequence:

input->length->test
input->idx0->test
input->idx1->test

we always construct the following dag:
 
input->length->test
  when true: input->idx0->test
      when true: input->idx1->test
       
which represents an indirect dependency on the length evaluation, therefore, it's garanteed to precede the indexer access in the ordered set.

Even in a multi-arm match we may use an earlier length temp from another pattern, but it won't get out of order in a sense that we could reach an uninitialized temp.

Does that address your concern?

Reply:
> Even in a multi-arm match we may use an earlier length temp from another pattern, but it won't get out of order in a sense that we could reach an uninitialized temp.

Right, that's what I was concerned about. Your explanation makes sense.
Basically, we only get to the indexing tests if the length test succeeds. The way we do that is that we stack the length test first in `MakeTestsAndBindingsForLengthAndListPatterns`. As a result we can rely on it having been computed already (the temp is initialized) by the time we get to the subsequent tests we stacked.